src/python/gudhi/wasserstein.py

```diff
@@ -62,36 +62,79 @@ def _perstot(X, order, internal_p):
+               new_match.append((-1, j))
+        elif j >= m:
+            if i < n:
+                new_match.append((i,-1))
+        else:
+            new_match.append((i,j))
+    return new_match
+
+
+def wasserstein_distance(X, Y, matching=False, order=2., internal_p=2.):
+    '''
+    :param X: (n x 2) numpy.array encoding the (finite points of the) first diagram. Must not contain essential points 
+                (i.e. with infinite coordinate).
     :param Y: (m x 2) numpy.array encoding the second diagram.
+    :param matching: if True, computes and returns the optimal matching between X and Y, encoded as...
     :param order: exponent for Wasserstein; Default value is 2.
-    :param internal_p: Ground metric on the (upper-half) plane (i.e. norm L^p in R^2); Default value is 2 (Euclidean norm).
-    :returns: the Wasserstein distance of order q (1 <= q < infinity) between persistence diagrams with respect to the internal_p-norm as ground metric.
-    :rtype: float
+    :param internal_p: Ground metric on the (upper-half) plane (i.e. norm L^p in R^2); 
+                        Default value is 2 (Euclidean norm).
+    :returns: the Wasserstein distance of order q (1 <= q < infinity) between persistence diagrams with 
+                        respect to the internal_p-norm as ground metric.
+              If matching is set to True, also returns the optimal matching between X and Y.
     '''
     n = len(X)
     m = len(Y)
 
     # handle empty diagrams
     if X.size == 0:
         if Y.size == 0:
-            return 0.
+            if not matching:
+                return 0.
+            else:
+                return 0., []
         else:
-            return _perstot(Y, order, internal_p)
+            if not matching:
+                return _perstot(Y, order, internal_p)
+            else:
+                return _perstot(Y, order, internal_p), [(-1, j) for j in range(m)]
     elif Y.size == 0:
-        return _perstot(X, order, internal_p)
+        if not matching:
+            return _perstot(X, order, internal_p)
+        else:
+            return _perstot(X, order, internal_p), [(i, -1) for i in range(n)]
 
     M = _build_dist_matrix(X, Y, order=order, internal_p=internal_p)
-    a = np.full(n+1, 1. / (n + m) )  # weight vector of the input diagram. Uniform here.
-    a[-1] = a[-1] * m                # normalized so that we have a probability measure, required by POT
-    b = np.full(m+1, 1. / (n + m) )  # weight vector of the input diagram. Uniform here.
-    b[-1] = b[-1] * n                # so that we have a probability measure, required by POT
+    a = np.ones(n+1) # weight vector of the input diagram. Uniform here.
+    a[-1] = m
+    b = np.ones(m+1) # weight vector of the input diagram. Uniform here.
+    b[-1] = n
+
+    if matching:
+        P = ot.emd(a=a,b=b,M=M, numItermax=2000000)
+        ot_cost = np.sum(np.multiply(P,M))
+        P[P < 0.5] = 0  # trick to avoid numerical issue, could it be improved?
```

Comment:
Do you still need this now that you are passing integer weights?

Reply:
Well, I did some trials and it seems the minimum is always `1.` so I think could remove this ; however I don't know how to be 100% sure that I won't have some 1E-10 numerical issue happening that could lead to weird bugs in further steps.

How bad is it to left this?

Reply:
From an algorithmic point of view, this method does not invent weights, it takes the weight of a point and adds/subtracts it to other weights. Integers should then remain integers quite safely. Actually, by default, the C++ code used by POT uses integral weights, and POT explicitly tells it to use double instead.
One issue with keeping the rounding is that if there are indeed numerical issues, we will never learn of it. And reciprocally, if the line is useless, we will not know, so we can never remove it.
If you really want to keep it, we can (I won't block the PR for that), although I'd rather replace it with `assert (P[P<1]==0).all()`.

Reply:
I understand your point ; I will remove it.